### ISSUE TITLE
Decide whether we want to standardize which `assert_eq!` parameter ("left" or "right")  should be used for "expected" value

### ISSUE THREAD

**Issue by findepi** [apache/datafusion#11270]
`assert_eq!` names its parameters "left" and "right" and doesn't seem to designate which one is expected and which is actual. 
However, when a test is run from RustRover, the IDE is apparently opinionated and currently expects the "right" to be "the expected" (see screenshot below).

Let's decide whether we as the project want to care about this. Currently, the code varies: [example 1](https://github.com/apache/datafusion/blob/fdd1e3db71c07f412366912396f51520147f3a10/datafusion/optimizer/tests/optimizer_integration.rs#L294), [example 2](https://github.com/apache/datafusion/blob/18042fd69138e19613844580408a71a200ea6caa/datafusion/optimizer/src/test/mod.rs#L227).

<img width="1156" alt="image" src="https://github.com/apache/datafusion/assets/144328/989d9c55-f032-49f0-880f-4bc93f672604">

Some RustRover references 

- https://github.com/intellij-rust/intellij-rust/issues/3814
- https://github.com/intellij-rust/intellij-rust/pull/3848

**Comment by findepi** [comment]
cc @alamb @jonahgao

**Comment by alamb** [comment]
I think consistency is a good thing. I personally prefer `assert(actual, expected)` but would follow whatever convention is chosen

According to the internet there appears to be no consensus across rust https://users.rust-lang.org/t/assert-eq-expected-and-actual/20304 (perhaps the `left` and `right` is a compromise between different opinions)

If we want to standardize (which seems like a good thing to me) it seems like we would:
1. Get (lazy) consensus
2. Add a note to the contributors guide: https://datafusion.apache.org/contributor-guide/index.html
3. Make as much of the code consistent as possible

**Comment by alamb** [comment]
(Thank you for raising this @findepi )

**Comment by Omega359** [comment]
FYI, somewhat related I do think RustRover has a bug around diffing results where it flips the expected and actual in the diff. See https://youtrack.jetbrains.com/issue/IJPL-157945. I need to provide a minimal test case for that

**Comment by findepi** [comment]
I agree @Omega359 this looks more like RustRover issue. It seems that RR has some limitations under which they are working (https://github.com/intellij-rust/intellij-rust/pull/3848).
At the same time, confusion about what's expected and what's actual isn't helping anyone (but this only matters when a test fails).

I came across an assertions library for Rust: https://github.com/google/assertor.
I like single easy entry point they have (`assert_that!`), with code completion for the assertions themselves.
They also attempt to provide useful err messages.
For example `assert_that!(vec![2, 4, 6]).contains_exactly(vec![2, 5, 6]);`  prints
```
assertion failed: datafusion/core/src/myfile.rs:88:9
missing (1)   : [5]
unexpected (1): [4]
```

Does anyone have experiences with it in a larger code base?

**Comment by Omega359** [comment]
That seems similar to the assertJ Java assertion library which I've used in the past. Fluent api's like that can be more verbose but in general I find them a lot easier to use.
